# Atividade Prática — QA com APIs (SWAPI)

## Versão do aluno — com scripts base

**Tempo estimado:** 60 minutos  
**Ferramenta:** Google Colab  
**Formato:** em grupos de **2 a 4 componentes** ou, se preferirem, **individualmente**.  
**Objetivo:** investigar respostas de uma API, ajustar scripts simples e responder às perguntas como um QA.

### Instruções
Nesta atividade, vocês **não precisam criar tudo do zero**.  
Os scripts já estão prontos como base. A tarefa do grupo será:

1. **Executar** os blocos de código
2. **Alterar apenas o que for pedido**
3. **Observar os resultados**
4. **Responder às perguntas em Markdown**
5. **Discutir em grupo** quais validações um QA faria

### Entrega
Cada grupo pode entregar **um único notebook** com:
- nomes dos integrantes
- respostas das perguntas
- scripts ajustados
- asserts completados

Se alguém optar por fazer sozinho, pode entregar individualmente sem problema.


In [223]:
import requests
import json

# -------------------------
# MODO DE SEGURANÇA (fallback)
# -------------------------
# Se a API não responder, o notebook usa dados locais
# para a atividade continuar funcionando.

DADOS_FALLBACK = {
    "https://swapi.dev/api/people/10/": {
        "name": "Obi-Wan Kenobi",
        "height": "182",
        "mass": "77",
        "eye_color": "blue-gray",
        "gender": "male",
        "url": "https://swapi.dev/api/people/10/"
    },
    "https://swapi.dev/api/people/13/": {
        "name": "Chewbacca",
        "height": "228",
        "mass": "112",
        "eye_color": "blue",
        "gender": "male",
        "url": "https://swapi.dev/api/people/13/"
    },
    "https://swapi.dev/api/planets/2/": {
        "name": "Alderaan",
        "climate": "temperate",
        "terrain": "grasslands, mountains",
        "population": "2000000000",
        "url": "https://swapi.dev/api/planets/2/"
    },
    "https://swapi.dev/api/starships/10/": {
        "name": "Millennium Falcon",
        "model": "YT-1300 light freighter",
        "manufacturer": "Corellian Engineering Corporation",
        "max_atmosphering_speed": "1050",
        "url": "https://swapi.dev/api/starships/10/"
    }
}

class RespostaLocal:
    def __init__(self, status_code, data):
        self.status_code = status_code
        self._data = data
        self.text = json.dumps(data, ensure_ascii=False)

    def json(self):
        return self._data

def buscar_endpoint(url):
    try:
        resposta = requests.get(url, timeout=10)
        if resposta.status_code == 200:
            return resposta, "online"
    except Exception:
        pass

    if url in DADOS_FALLBACK:
        return RespostaLocal(200, DADOS_FALLBACK[url]), "fallback local"

    return RespostaLocal(404, {"detail": "Not found"}), "fallback local"

print("Ambiente preparado com sucesso.")

Ambiente preparado com sucesso.


## Identificação do grupo

- Integrante 1: Kauan Borges Araujo
- Integrante 2:
- Integrante 3 (se houver):
- Integrante 4 (se houver):

## Parte 1 — Explorando a API

Nesta primeira parte, vocês vão consultar 4 endpoints.  
O trabalho do grupo será **alterar apenas a URL** em cada bloco, executar e observar o retorno.


### 1.1 Personagem 11

**Tarefa:** execute o script e observe:
- status code
- nome
- altura
- cor dos olhos

In [224]:
url = "https://swapi.dev/api/people/11/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Anakin Skywalker",
    "height": "188",
    "mass": "84",
    "hair_color": "blond",
    "skin_color": "fair",
    "eye_color": "blue",
    "birth_year": "41.9BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/1/",
    "films": [
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [],
    "vehicles": [
        "https://swapi.dev/api/vehicles/44/",
        "https://swapi.dev/api/vehicles/46/"
    ],
    "starships": [
        "https://swapi.dev/api/starships/39/",
        "https://swapi.dev/api/starships/59/",
        "https://swapi.dev/api/starships/65/"
    ],
    "created": "2014-12-10T16:20:44.310000Z",
    "edited": "2014-12-20T21:17:50.327000Z",
    "url": "https://swapi.dev/api/people/11/"
}


**Respostas do aluno — Personagem 11**

- Status code: **200**
- Nome do personagem: **Anakin Skywalker**
- Altura: **188cm**
- Cor dos olhos: **Azul**
- O JSON parece completo? **Sim**

### 1.2 Personagem 20

**Tarefa:** neste bloco, troque apenas o endpoint se necessário, execute e observe o retorno.

In [225]:
url = "https://swapi.dev/api/people/20/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Yoda",
    "height": "66",
    "mass": "17",
    "hair_color": "white",
    "skin_color": "green",
    "eye_color": "brown",
    "birth_year": "896BBY",
    "gender": "male",
    "homeworld": "https://swapi.dev/api/planets/28/",
    "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/",
        "https://swapi.dev/api/films/4/",
        "https://swapi.dev/api/films/5/",
        "https://swapi.dev/api/films/6/"
    ],
    "species": [
        "https://swapi.dev/api/species/6/"
    ],
    "vehicles": [],
    "starships": [],
    "created": "2014-12-15T12:26:01.042000Z",
    "edited": "2014-12-20T21:17:50.345000Z",
    "url": "https://swapi.dev/api/people/20/"
}


**Respostas do aluno — Personagem 20**

- Status code: **200**
- Nome do personagem: **Yoda**
- Altura: **66cm**
- Cor dos olhos: **Marrom**
- O JSON parece completo? **Sim**

### 1.3 Planeta 7

**Tarefa:** execute o bloco e observe:
- nome do planeta
- clima
- terreno
- população

In [226]:
url = "https://swapi.dev/api/planets/7/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Endor",
    "rotation_period": "18",
    "orbital_period": "402",
    "diameter": "4900",
    "climate": "temperate",
    "gravity": "0.85 standard",
    "terrain": "forests, mountains, lakes",
    "surface_water": "8",
    "population": "30000000",
    "residents": [
        "https://swapi.dev/api/people/30/"
    ],
    "films": [
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-10T11:50:29.349000Z",
    "edited": "2014-12-20T20:58:18.429000Z",
    "url": "https://swapi.dev/api/planets/7/"
}


**Respostas do aluno — Planeta 7**

- Status code: **200**
- Nome do planeta: **Endor**
- Clima: **Temperado**
- Terreno: **Florestas, Montanhas e Lagos**
- População: **30000000**
- O JSON parece completo? **Sim**

### 1.4 Nave 17

**Tarefa:** execute o bloco e observe:
- nome da nave
- modelo
- fabricante
- velocidade máxima em atmosfera

In [227]:
url = "https://swapi.dev/api/starships/17/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print(json.dumps(dados, indent=4, ensure_ascii=False))

Origem dos dados: online
Status Code: 200
{
    "name": "Rebel transport",
    "model": "GR-75 medium transport",
    "manufacturer": "Gallofree Yards, Inc.",
    "cost_in_credits": "unknown",
    "length": "90",
    "max_atmosphering_speed": "650",
    "crew": "6",
    "passengers": "90",
    "cargo_capacity": "19000000",
    "consumables": "6 months",
    "hyperdrive_rating": "4.0",
    "MGLT": "20",
    "starship_class": "Medium transport",
    "pilots": [],
    "films": [
        "https://swapi.dev/api/films/2/",
        "https://swapi.dev/api/films/3/"
    ],
    "created": "2014-12-15T12:34:52.264000Z",
    "edited": "2014-12-20T21:23:49.895000Z",
    "url": "https://swapi.dev/api/starships/17/"
}


**Respostas do aluno — Nave 17**

- Status code: **200**
- Nome da nave: **Rebel transport**
- Modelo: **GR-75 medium transport**
- Fabricante: **Gallofree Yards, Inc**
- Velocidade máxima em atmosfera: **650**
- O JSON parece completo? **Sim**

## Parte 2 — Extraindo campos específicos

Agora você vai executar os scripts e **completar as respostas**.

### 2.1 Extraindo campos do personagem

**Tarefa:** execute o bloco e depois responda.

In [228]:
url = "https://swapi.dev/api/people/43/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome:", dados["name"])
print("Altura:", dados["height"])
print("Cor dos olhos:", dados["eye_color"])
print("Gênero:", dados["gender"])

Nome: Shmi Skywalker
Altura: 163
Cor dos olhos: brown
Gênero: female


**Perguntas**
1. Qual campo foi usado para mostrar o nome? **Campo 'name'**
2. Qual campo foi usado para mostrar a altura? **Campo 'height'**
3. Se você fosse QA, quais desses campos validaria primeiro?

**R: Validaria primeiro o 'name' pois além do URL, busco a instrução do personagem através dele.**

### 2.2 Extraindo campos do planeta

**Tarefa:** execute o bloco e complete as respostas.

In [236]:
url = "https://swapi.dev/api/planets/11/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Nome do planeta:", dados["name"])
print("Clima:", dados["climate"])
print("Terreno:", dados["terrain"])
print("População:", dados["population"])

Nome do planeta: Geonosis
Clima: temperate, arid
Terreno: rock, desert, mountain, barren
População: 100000000000


**Perguntas**
1. O campo população parece numérico? **Sim, o campo é numérico**
2. O clima foi retornado corretamente? **Sim**
3. Se você fosse QA, o que validaria nesse planeta?

**R: Acredito que assim como o personagem, validaria primeiro o 'name' para evitar conflito de informações iguais e a população por distoar o tipo primitivo do restante dos dados do planeta, que são Strings.**

## Parte 3 — Pensando como QA

Agora você vai trabalhar com validações simples.

### 3.1 Validações básicas

**Tarefa:** execute o bloco abaixo e depois escreva o que cada validação verifica.

In [237]:
url = "https://swapi.dev/api/people/10/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Validação 1 - status 200:", response.status_code == 200)
print("Validação 2 - campo 'name' existe:", "name" in dados)
print("Validação 3 - nome não está vazio:", dados["name"] != "")
print("Validação 4 - campo 'height' existe:", "height" in dados)

Validação 1 - status 200: True
Validação 2 - campo 'name' existe: True
Validação 3 - nome não está vazio: True
Validação 4 - campo 'height' existe: True


**Explique com suas palavras**
1. O que a validação 1 verifica? **Verifica o Status Code da informação buscada.**
2. O que a validação 2 verifica? **Verifica se o campo 'name' existe.**
3. O que a validação 3 verifica? **Verifica se o nome não está vazio**
4. O que a validação 4 verifica? **Verifica se o campo 'height' existe.**

### 3.2 Complete as validações

**Tarefa:** execute o bloco abaixo. Ele já está funcionando. Depois, alterem as expressões se o professor pedir outra nave ou outro endpoint. Em seguida, expliquem por que cada validação faz sentido para QA.


In [246]:
url = "https://swapi.dev/api/starships/5/"
response, origem = buscar_endpoint(url)
dados = response.json()

print("Status 200:", response.status_code == 200)

# Os exemplos abaixo já estão corretos e executáveis.
# Se o professor pedir, vocês podem alterar para validar outros campos.
print("Campo 'name' existe:", "name" in dados)
print("Campo 'model' existe:", "model" in dados)
print("Nome não está vazio:", dados["name"] != "")


Status 200: True
Campo 'name' existe: True
Campo 'model' existe: True
Nome não está vazio: True


## Parte 4 — Testes automatizados com assert

Agora vocês vão trabalhar sobre scripts já prontos. Os asserts abaixo já executam sem erro. A tarefa do grupo é **entender o que cada linha valida**, ajustar se o professor pedir outro recurso e registrar as respostas.


### 4.1 Assert no personagem

**Tarefa:** execute o bloco. Depois, expliquem o que cada assert valida e, se quiserem, alterem o endpoint para outro personagem.


In [248]:
url = "https://swapi.dev/api/people/19/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "name" in dados, "Erro: campo name não encontrado"
assert dados["name"] != "", "Erro: nome vazio"

print("Todos os testes do personagem passaram!")


Todos os testes do personagem passaram!


**R: A primeira linha de assert valida se o personagem existe, retornando Status code 200 caso a resposta seja 'true', se não, retorna "Erro: status inesperado". A segunda linha de assert retorna o campo nome do personagem, se não, retorna "Erro: Campo name não encontrado". A terceira linha de assert retorna o nome do personagem se ele for diferente de campo vazio, se não retorna "Erro: nome vazio".**

### 4.2 Assert na nave

**Tarefa:** execute o bloco. Depois, descrevam qual campo foi validado em cada assert e proponham mais 1 validação extra em Markdown.


In [249]:
url = "https://swapi.dev/api/starships/15/"
response, origem = buscar_endpoint(url)
dados = response.json()

assert response.status_code == 200, "Erro: status inesperado"
assert "manufacturer" in dados, "Erro: fabricante não encontrado"
assert dados["model"] != "", "Erro: modelo vazio"

print("Todos os testes da nave passaram!")


Todos os testes da nave passaram!


**R: A primeira linha de assert valida se a nave existe, retornando Status code 200 caso a resposta seja 'true', se não, retorna "Erro: status inesperado". A segunda linha de assert retorna o campo fabricante da nave, se não, retorna "Erro: fabricante não encontrado". A terceira linha de assert retorna o modelo se ele for diferente de campo vazio, se não, retorna "Erro: modelo vazio".**

## Parte 5 — Cenário negativo

Agora você vai observar como a API se comporta quando pedimos um recurso inexistente.

In [234]:
url = "https://swapi.dev/api/people/9999/"
response, origem = buscar_endpoint(url)

print("Origem dos dados:", origem)
print("Status Code:", response.status_code)
print("Texto retornado:")
print(response.text)

Origem dos dados: fallback local
Status Code: 404
Texto retornado:
{"detail": "Not found"}


**Perguntas**
1. Qual status code foi retornado? **404**
2. Esse comportamento está correto? **Sim, visto que não temos o endpoint 9999, é correto o programa nos retornar o '404 Not Found'.**
3. Por que um QA deve testar cenários negativos? **Para verificar se o programa possui erros que possam comprometer a experiência do usuário.**

## Parte 6 — Mini desafio final

Agora você vai alterar apenas o endpoint da função abaixo e gerar um pequeno resumo.

In [256]:
def resumo_recurso(url):
    response, origem = buscar_endpoint(url)
    print("Origem dos dados:", origem)
    print("Status Code:", response.status_code)

    if response.status_code != 200:
        print("Recurso não encontrado.")
        return

    dados = response.json()

    for chave, valor in dados.items():
        print(f"{chave}: {valor}")

# ALTERE O ENDPOINT ABAIXO PARA TESTAR OUTRO RECURSO
resumo_recurso("https://swapi.dev/api/planets/3/")

Origem dos dados: online
Status Code: 200
name: Yavin IV
rotation_period: 24
orbital_period: 4818
diameter: 10200
climate: temperate, tropical
gravity: 1 standard
terrain: jungle, rainforests
surface_water: 8
population: 1000
residents: []
films: ['https://swapi.dev/api/films/1/']
created: 2014-12-10T11:37:19.144000Z
edited: 2014-12-20T20:58:18.421000Z
url: https://swapi.dev/api/planets/3/


**Respostas finais**
1. Qual endpoint você testou no desafio? **3**
2. Qual foi o status code? **200**
3. Qual campo chamou mais a sua atenção? **'Population' por ser muito pequena.**
4. Se você fosse QA, quais 3 validações faria nesse recurso? **'Name' para me instruir sobre os dados que acabo de buscar além do próprio endpoint, 'films' por trabalhar com outra requisição em API e o campo 'residents' que também busca requisição em API porém ao utilizar esse endpoint, não houve requisições, deixando o campo vazio.**




# Entrega da atividade

Entreguem:

1. O notebook preenchido  
2. As respostas em Markdown  
3. Os scripts ajustados  
4. Os asserts analisados e/ou adaptados  
5. Uma conclusão curta do grupo: **quais foram as 3 validações mais importantes?**

## Fechamento
Ao final desta atividade, vocês praticaram:

- leitura de resposta JSON
- análise de status code
- validação de campos
- pensamento de QA
- testes simples com assert
- análise de cenário negativo
